### PR TITLE
Add merge method to Component

### PR DESCRIPTION
**Pull request by hsaunders1904**
## Linked Issues

Partially addresses #892.

## Description

Adds a `merge_children` method to components so that we can easily merge component trees within builds/designs.

This can be particularly helpful when building subcomponents of another component in multiple dimensions. This new method can be used when the xy, xz, and xyz components of each subcomponent need to be merged.

## Interface Changes

Added `merge_children` method to `Component`.

## Checklist

I confirm that I have completed the following checks:

- [ ] Tests run locally and pass `pytest tests --reactor`
- [x] Code quality checks run locally and pass `flake8` and `black .`
- [x] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

**Review by CoronelBuendia — APPROVED** [review]

File: bluemira/base/components.py [hunk 1]
```diff
@@ -158,6 +207,12 @@ def add_children(self, children: List[Component], merge_trees=False):
...
             This component.
         """
+        if isinstance(children, list) and len(children) == 0:
+            # I'm not sure what the block of code below this is for, but
+            # I assume it does something useful. However, it will throw
+            # an IndexError on an empty list, avoid that here.
+            return self
+
         if not isinstance(children, list) or len(children) == 0:
             child = children[0] if isinstance(children, list) else children
```
Comment (je-cook) [inline bluemira/base/components.py:217]: any chance of a comment here @CoronelBuendia git blame tells me its apparently yours :slightly_smiling_face:?

Probably should narrow it down so we don't say "I dont know what this does" in the code

**Review by je-cook — CHANGES_REQUESTED** [review]
looks good to me. 

One question, how is the connection between leaves reconstructed eg if you want `xy` and `leaf_x_2` is the x that is required? Is it just implicit because of the order the merge is constructed? 

If that doesn't make sense let me know and I will reword better

File: bluemira/base/components.py (hunk 1 above)
Comment (CoronelBuendia) [reply bluemira/base/components.py:217]: :O That doesn't look like me... I don't normally write 1 list if/elses... but if git blame says so, then I guess I did? If @DanShort12 has no recollection and the tests pass without it, I'd say get rid of anything we don't understand

**Comment by hsaunders1904** [comment]
> @je-cook  One question, how is the connection between leaves reconstructed eg if you want xy and leaf_x_2 is the x that is required? Is it just implicit because of the order the merge is constructed?

Hmm, yeah, it's implicit. I think the behaviour for this is probably wrong.  An error may be more consistent, seeing as we raise if there are common non-leaf components. What do you think?

**Comment by hsaunders1904** [comment]
Reverted [920a32a](https://github.com/Fusion-Power-Plant-Framework/bluemira/pull/933/commits/920a32a38ecd7b09d225fa0b92129d85eee3ff00) in favour of opening a separate ticket: https://github.com/Fusion-Power-Plant-Framework/bluemira/issues/955.

**Comment by je-cook** [comment]
The last commit says TODO is this still waiting on some changes?

**Pull request closed without merging**